About query optimization and considerations for scaling (#182)
crate/cratedb-guide: docs/performance/productizing.rst

Reviewer: Even if it's one more nitpick, I think "productizing", as used in the filename, is the wrong word.

However, I don't want to bother you too much about it: As long as you don't quickly spread the word about the new resource URL, I can easily run a follow-up patch for you to review that adjusts a bit more things on the "naming things" details if you don't mind.

Reviewer: What about calling the file just `scaling.rst`, and use an imperative but concise "Design for scaling" as a title? Or would "Design for scale" be better/correct?

.. _topics-to-watch-out-when-productizing-cratedb:

###############################################
 Topics to watch out when productizing CrateDB

> ###############################################

Reviewer: That's a bit of an outlier. Can you try to find a more concise headline and link label?

![image](https://github.com/user-attachments/assets/160b2c47-6986-4da4-90a5-ec49b24167c5) 
![image](https://github.com/user-attachments/assets/087cf9a0-6657-475f-935c-5fc4bb39e0f8)

Reviewer: It can be made different on the navigation vs. on the page headline, when needed. In this case, I think both spots can benefit.

Reviewer: Is "productizing" meant in the same sense like [Going into production](https://cratedb-guide--182.org.readthedocs.build/admin/going-into-production.html), but focusing on performance details instead of operational ones?

Reviewer: What about a possible shorter "Optimize for production" in this case, or, alternatively, a more general "Production notes"?

Author: I changed it to the shorter "Key design considerations for scaling" now, but it seems it is still too long?

Reviewer: It's still a bit long, yes. But we can also change it after an initial merge.

.. _mindful-of-memory:

*******************************
 Be mindful of memory capacity
*******************************

> In CrateDB operations requiring a working set like groupings, aggregations, and

Reviewer 2: ```suggestion
In CrateDB, operations requiring a working set like groupings, aggregations, and
```

sorting are performed fully in memory without spilling over to disk.

Sometimes you may have a query that leads to a sub-optimal execution plan
requiring lots of memory. If you are coming to CrateDB from other database
systems, your experience may be that these queries will proceed to run taking

> longer than required and impacting other workloads in the meanwhile, sometimes

Reviewer 2: ```suggestion
longer than required and impacting other workloads in the meanwhile. Sometimes
```
To make sentences a bit shorter.

this effect may be obvious if a query takes a lot of resources and runs for a
long time, other times it may go unnoticed if a query that could complete in say
100 milliseconds takes one hundred times longer, 10 seconds, but the users put
up with it without reporting to you.

In CrateDB if a query would require more heap memory than the interested nodes
have available the query will fail with a particular type of error message that

> we call a `CircuitBreakerException`, this is a fail-fast approach, we can see

Reviewer 3: Why is this a fail-fast approach? It fails once the memory is occupied at runtime, could happen very late while processing queries. The main point here is: preventing the node from dying with an OOM.

Author: > It fails once the memory is occupied at runtime

In some cases yes, but in other cases I believe it does an estimate and it returns with the `CircuitBreakerException` right away based on the estimate, maybe you could check this?

Reviewer 3: Estimates can only be done once the data is processed. When using e.g. literals, the value is available already and the estimate can be done immediately.

Reviewer 3: In what cases did you experience "immediate" CBE's?

quickly that there is an issue and optimize the query to get the best
performance, without impacting other workloads.

Please take a look at :ref:`Query Optimization 101 <performance-optimization>` for strategies to optimize your
queries when you encounter this situation.

.. _reading-lots-of-records:

*************************
 Reading lots of records
*************************

> When the http endpoint is used CrateDB will prepare the entire response in

Reviewer 2: ```suggestion
When the HTTP endpoint is used CrateDB will prepare the entire response in
```

memory before sending it to the client.

When the PostgreSQL protocol is used CrateDB attempts to stream the results but

> in many cases it still needs to bring all rows to the query handler node first.

Reviewer 3: Why *attempts*? With PG, it will always stream data. Also inter-node data is paged, depending on the query (e.g the handler node does not have to apply any pipeline breaker like an final in-memory ordering), the data is also paged directly to the client. Thus, the handler node does not need to hold the complete result set in-memory in most cases.

> So we should always limit how many rows we request at a time, see `Fetching
> large result sets from CrateDB`_

Reviewer 2: ```suggestion
So we should always limit how many rows we request at a time, see `Fetching
large result sets from CrateDB`_.
```

.. _number-of=shards:

******************
 Number of shards
******************

In CrateDB data in tables and partitions is distributed in storage units that we
call shards.

> If we do not specify how many shards we want for a table/partition CrateDB will
> derive a default from the number of CPUs.

Reviewer 2: ```suggestion
If we do not specify how many shards we want for a table/partition CrateDB will
derive a default from the number of nodes.
```

https://cratedb.com/docs/crate/reference/en/latest/sql/statements/create-table.html#clustered:
> num_shards = max(4, num_data_nodes * 2)

> derive a default from the number of nodes.

Reviewer 3: The default is **4**, it will **not** derive it from the number of nodes.

Author: https://cratedb.com/docs/crate/reference/en/latest/sql/statements/create-table.html#clustered
says:

> If not provided, the number of shards is calculated based on the number of currently active data nodes

Reviewer 3: Oh right, my fault! 🤦  I should probably develop more often with > 4 nodes ;)

CrateDB also has replicas of data and this results in additional shards in the
cluster.

Having too many or too few shards has performance implications, so it is very

> important to get familiar with the :ref:`Sharding Performance Guide <sharding_guide>`

Reviewer 2: ```suggestion
important to get familiar with the :ref:`Sharding Performance Guide <sharding_guide>`.
```

But in particular, there is a soft limit of 1000 shards per node; so table
schemas, partitioning strategy, and number of nodes need to be planned to stay
well below this limit, one strategy can be to aim for a configuration where even
if one node in the cluster is lost the remaining nodes would still have less
than 1000 shards.

If this was not considered when initially defining the tables we have the
following considerations:

-  changing the partitioning strategy requires creating a new table and copying
   over the data
-  the easiest way to change the number of shards on a partitioned table is to
   do it for new shards only with the `ALTER TABLE ONLY` command
-  see also `Changing the number of shards`_

.. _amount-of-indexed-columns:

*****************************
 Amount of (indexed) columns
*****************************

CrateDB indexes everything by default, providing excellent query performance
without requiring manual indexing. However, excessive indexing can impact
storage, write speed, and resource utilization.

-  All columns are automatically indexed when inserted.
-  CrateDB optimizes indexing using Lucene-based columnar storage.

> -  A soft limit of 1,000 indexed columns (or JSON fields) per table exists.

Author: You are right, this does not seem to apply to first level columns:
```sql
CREATE TABLE t1 (a int);
    ALTER TABLE t1 SET ("mapping.total_fields.limit"=3);
    ALTER TABLE t1 ADD column b int;
    ALTER TABLE t1 ADD column c int;
    ALTER TABLE t1 ADD column d int INDEX OFF STORAGE WITH (columnstore = false);
--> SQLParseException[Limit of total columns [3] in table [doc.t1] exceeded]
```

This paragraph was originally from @wierdvanderhaar , we should edit it to make it about `OBJECT`s.

> -  A soft limit of 1,000 indexed columns (or JSON fields) per table exists.

Author: I updated this section now, please review if you have a chance.

-  Going beyond this limit may impact performance.

In cases with many fields and columns, it is advised to determine which columns
need to be indexed.

.. _section-joins:

*******
 JOINs
*******

CrateDB is a lot better at JOINs than many of our competitors and is getting
better at every release, but JOINs in distributed databases are tricky to

> optimize, so in many cases queries involving JOINs may need a bit of tweaking.

Reviewer: I get the message, and I love it. However, I think the sound is a bit sloppy, specifically the bragging fragment

> is a lot better at JOINs than many of our competitors

even if it's true. Maybe we can find a better wording? /cc @kneth, @simonprickett

Reviewer: Maybe be more specific, possibly even to the extent of doing name dropping, like...

> CrateDB's engine to coordinate and execute distributed JOINs is already more efficient
> than many competing implementations {by other vendors | like ArangoDB, Cassandra, CockroachDB, MongoDB}, building upon years of maturing since its introduction with CrateDB 1.0.
> Because JOIN operations in distributed databases are tricky to optimize, 
> they need a bit of query tweaking, based on what the query planner says.

So, I've added another variant of bragging about CrateDB's implementation ("building upon years of maturing since its introduction with CrateDB 1.0"), and also another variant of sloppyness ("based on what the query planner says"). I am sure you will make the best out of it?

.. _topics-to-watch-out-when-productizing-cratedb:

#######################################
 Key design considerations for scaling
#######################################

This article explores critical design considerations to successfully scale
CrateDB in large production environments to ensure performance and reliability
as workloads grow.

.. _mindful-of-memory:

*******************************
 Be mindful of memory capacity
*******************************

In CrateDB, operations requiring a working set like groupings, aggregations, and
sorting are performed fully in memory without spilling over to disk.

Sometimes you may have a query that leads to a sub-optimal execution plan
requiring lots of memory. If you are coming to CrateDB from other database
systems, your experience may be that these queries will proceed to run taking
longer than required and impacting other workloads in the meanwhile. Sometimes
this effect may be obvious if a query takes a lot of resources and runs for a
long time, other times it may go unnoticed if a query that could complete in say
100 milliseconds takes one hundred times longer, 10 seconds, but the users put
up with it without reporting to you.

> In CrateDB if a query would require more heap memory than the interested nodes

Reviewer 4: ```suggestion
If a query would require more heap memory than the interested nodes
```

have available the query will fail with a particular type of error message that

> we call a ``CircuitBreakerException``, this is a fail-fast approach, we can see

Reviewer 4: ```suggestion
we call a ``CircuitBreakerException``. This is a fail-fast approach as we quickly see
```

> quickly that there is an issue and optimize the query to get the best

Reviewer 4: ```suggestion that there is a performance issue and optimize the query to get the best
```
